crates/cubecl-core/src/ir/operation.rs

```diff
@@ -370,33 +370,39 @@ pub struct FmaOperator {
 }
 
 #[allow(missing_docs)]
-pub struct CheckedIndexAssign {
-    pub lhs: Variable,
-    pub rhs: Variable,
-    pub out: Variable,
-}
-
-impl CheckedIndexAssign {
-    #[allow(missing_docs)]
-    pub fn expand(self, scope: &mut Scope) {
-        let lhs = self.lhs;
-        let rhs = self.rhs;
-        let out = self.out;
-        let array_len = scope.create_local(Item::new(Elem::UInt(UIntKind::U32)));
-        let inside_bound = scope.create_local(Item::new(Elem::Bool));
-
-        if out.has_buffer_length() {
-            cpa!(scope, array_len = buffer_len(out));
-        } else {
-            cpa!(scope, array_len = len(out));
-        }
+pub fn expand_checked_index(scope: &mut Scope, lhs: Variable, rhs: Variable, out: Variable) {
+    let array_len = scope.create_local(Item::new(Elem::UInt(UIntKind::U32)));
+    let inside_bound = scope.create_local(Item::new(Elem::Bool));
+    let item = scope.create_local(out.item);
+    let zero: Variable = 0u32.into();
+
+    if lhs.has_buffer_length() {
+        cpa!(scope, array_len = buffer_len(lhs));
+    } else {
+        cpa!(scope, array_len = len(lhs));
+    }
 
-        cpa!(scope, inside_bound = lhs < array_len);
+    cpa!(scope, inside_bound = rhs < array_len);
 
-        cpa!(scope, if(inside_bound).then(|scope| {
-            cpa!(scope, unchecked(out[lhs]) = rhs);
-        }));
+    cpa!(scope, item = unchecked(lhs[rhs]));
+    cpa!(scope, out = select(inside_bound, item, zero));
+}
+
+#[allow(missing_docs)]
+pub fn expand_checked_index_assign(scope: &mut Scope, lhs: Variable, rhs: Variable, out: Variable) {
+    let array_len = scope.create_local(Item::new(Elem::UInt(UIntKind::U32)));
+    let inside_bound = scope.create_local(Item::new(Elem::Bool));
```

Comment:
Could use local_binding here (or in the next PR)

Reply:
Will be in the next PR since it leads to an error in this one.